Content plugin events

robbiejackson opened the pull request:
### **User description**
This is a draft structure for plugin events - submitted for review.

It's similar to the existing format at https://docs.joomla.org/Plugin/Events/Content but with some differences.

Let me know if you disagree with any of the following:

- the index page shows the event traditional name and concrete event class
- the index page shows when the event is available from. I know that it should be possible to infer this from selecting the version within docusaurus, but it's easy to make a mistake in the manual when duplicating files into different versions. Having the version there makes it clearer if a mistake has been made, and it's then easier to recover as well.
- the detailed descriptions are split into pages, based on the first segment of the concrete class (rather than the traditional name - the way it has been done up to now)
- the headers on the detail pages have the event class first - so there's a deliberate move to using these rather than the traditional names
- the examples are really just for examples which are properly documented in the Joomla manual. The additional notes can be used for other instances within Joomla, but these would require more work from the reader in understanding the Joomla code. 

Also, it might be worth at some point including sortable headers - as in https://pester.dev/docs/additional-resources/articles/


___

### **PR Type**
Documentation


___

### **Description**
- Added detailed documentation on accessing event arguments and returning values for different event classes in Joomla.
- Documented various content plugin events, including their descriptions, arguments, return values, and examples.
- Created an index page listing plugin events with links to their detailed descriptions.



___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>joomla-4-and-5-changes.md</strong><dd><code>Detailed Guide on Accessing Event Arguments and Returning Values</code></dd></summary>
<hr>

docs/building-extensions/plugins/joomla-4-and-5-changes.md

<li>Added detailed sections on accessing event arguments and returning <br>values for concrete, generic, and traditional legacy methods.<br> <li> Included code examples for each method.<br> <li> Explained the commitment to preserving argument order until Joomla 6.<br>


</details>


  </td>
  <td><a href="https://github.com/joomla/Manual/pull/291/files#diff-831bcd24dc10caf692ded68d0c8c598b81249f684d20bfe7cfb4996104940ef2">+147/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>content.md</strong><dd><code>Comprehensive Documentation of Content Plugin Events</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/building-extensions/plugins/plugin-events/content.md

<li>Documented various content plugin events with descriptions, arguments, <br>return values, and examples.<br> <li> Provided detailed information on events like onContentPrepare, <br>onContentAfterTitle, onContentBeforeDisplay, etc.<br> <li> Included references to Joomla manual and API documentation.<br>


</details>


  </td>
  <td><a href="https://github.com/joomla/Manual/pull/291/files#diff-80a6a68aebf8005ad466409c2436b0e684e22f0b9b04940243d15f4ae3c823ee">+451/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.md</strong><dd><code>Index of Plugin Events with Detailed Links</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/building-extensions/plugins/plugin-events/index.md

<li>Created an index page listing plugin events with their names, classes, <br>groups, and release versions.<br> <li> Added links to detailed descriptions of each event.<br> <li> Included a warning about coding for both concrete and generic event <br>classes.<br>


</details>


  </td>
  <td><a href="https://github.com/joomla/Manual/pull/291/files#diff-8ccca2a9725a7ae619693792968cbd9951ab65957d28840a597b3c6d6b120ea4">+27/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

HLeithner commented:
About the sortable datatables, pester (which you linked to) uses there own table implementation https://github.com/pester/docs/tree/main/src/components/PesterDataTable

Fedik commented on docs/building-extensions/plugins/plugin-events/content.md:
```diff
@@ -0,0 +1,140 @@
+
+- **`context`** - The context of the content being passed to the plugin. This is the component name and view - or name of module (e.g. com_content.article, com_contact.contact, com_users.user). Use this to check whether you are in the desired context for the plugin.
+
+- **`&item`** - A reference to the item which is being rendered by the view, for example, an article, contact or user. 
```
Just a note:
All this `referencing` will be need to update at some point later, we do not use it anymore.

Also whole `Parameters` section will be need to replace to "Event attributes", or something kind of that, similar to 
https://manual.joomla.org/docs/building-extensions/plugins/editors-xtd-plugin#oneditorbuttonssetup

I know it is copy from old page, so for now it is fine as it is, for beginning.

Fedik commented on docs/building-extensions/plugins/plugin-events/content.md:
```diff
@@ -0,0 +1,214 @@
+This is the first stage in preparing content for output and is the most common point for content-orientated plugins to do their work. 
+Since the item and related parameters are passed by reference, event handlers can modify them prior to display. 
+
+### Event Arguments
```
Sorry for confusing. 
The old parameters cannot be directly replaced with arguments. The naming is not always matches.
But we probably still need to keep both, because the old parameter names may be useful to help transition old plugins to event classes.

Example, old listener:
```php
function($context, $item, $params, $page = 0) {}
```

And new https://github.com/joomla/joomla-cms/blob/721c32cce32f6b6f64ec8b5a4ee5817e8a5b79e5/libraries/src/Event/Content/ContentPrepareEvent.php#L18-L24:
```php
// Event
$event = new ContentPrepareEvent('onContentPrepare', [
  'context' => 'com_example.example', 
  'subject' => $contentObject, 
  'params' => $params, 
  'page' => $pageNum
]);

// Listener
function(ContentPrepareEvent $event) {
  $context = $event->getContext(); // or $event->getArgument('context');
  $item = $event->getItem(); // or $event->getArgument('subject');
  $params = $event->getParams(); // or $event->getArgument('params');
  $page = $event->getPage(); // or $event->getArgument('page');
}
```

I curently not sure how to make it right. 
Maybe keep old "Parameters" section, and later add "Arguments section". And remove "Parameters" after Joomla 6 release. Or mix it somehow. Or maybe you have some better ideas :)

robbiejackson replied:
I was really following https://www.dionysopoulos.me/book/plg.html#plg-forms-j4-listener-types (see section in red below this).

I've used his suggested approach and found that it has worked ok. 

As he says in that section, even after Joomla uses a concrete event class for an event, you can still have other 3rd party components raising a generic event for that same event name, and Joomla will allow that, so your plugin has to support both. If Joomla is going to forbid that in Joomla 6 then we could move to just using concrete events at that stage. Do you know if it will?

I think that if the documentation matches the names of the arguments and the order of the arguments, then I'm hoping that this approach should work. Do you know anywhere specifically where it won't work?

Nik also mentioned some code he had submitted to ease this problem. Do you know what that is? Would including that into Joomla help things?

With regard to using Parameters, I'm reluctant to recommend a way of handling the events if we know it's going to fail in the near future (eg in Joomla 6).

Fedik replied:
Yes, you are right, it will still work in j6, I just meant "in future" :smiley:  

`CMSPlugin::registerLegacyListener()` is not deprecated, so cannot be removed in j6, and old listeners should still work there. For now only a plan to remove in j7 https://github.com/joomla/joomla-cms/pull/43395, but still under question.

> I think that if the documentation matches the names of the arguments and the order of the arguments, then I'm hoping that this approach should work.

A few things will not match for sure. Some events have a `subject` argument, which is kind of main "subject of event". Which in old event could be as `$form`, `$data`, `$item` etc. parameter.

However, I tried make an examples, mostly every event class have an example, kind of 
- https://github.com/joomla/joomla-cms/blob/721c32cce32f6b6f64ec8b5a4ee5817e8a5b79e5/libraries/src/Event/Content/ContentPrepareEvent.php#L18-L24
- https://github.com/joomla/joomla-cms/blob/721c32cce32f6b6f64ec8b5a4ee5817e8a5b79e5/libraries/src/Event/Model/AfterSaveEvent.php#L16-L20

Where it is possible to look actual arguments.

The defination for ordering, is like https://github.com/joomla/joomla-cms/blob/721c32cce32f6b6f64ec8b5a4ee5817e8a5b79e5/libraries/src/Event/Content/ContentPrepareEvent.php#L35
(it set to remove in j6, but I think it will be extended).



> Nik also mentioned some code he had submitted to ease this problem. Do you know what that is? Would including that into Joomla help things?

I do not remember, it probably was very long time ago :wink:

robbiejackson replied:
I'm coming to the conclusion that we have to explain the options to readers, and give them enough information so that they can choose what's best for them. So I've expanded the collapsible sections on how to access the parameters and return values. Would you mind checking that what I've written is correct?

I'm thinking also that we should maybe change 'item' to 'subject' in the arguments as well. What do you think? In the existing documentation it's described as 'article' which is definitely not accurate.

robbiejackson commented:
@HLeithner @Fedik many thanks for your comments - very useful.

I've revised the documentation
- events now split by plugin group - I presume the Model events will fit under Content group
- I've added how to access the event arguments and return a value on the Content plugins page, but as a collapsible section. I'm presuming we would add that on each page. The event arguments are going to be different for different plugin groups, so you can't have a generic section of text that would do them all.
- additional notes are now set as info admonitions
- examples section is included only if there are valid examples
- On the details page I've set the toc_max_heading_level to just display the event names

Details page is at 
http://pr-291.manual.joomlacode.org/docs/next/building-extensions/plugins/plugin-events/content

Fedik commented:
> So I've expanded the collapsible sections on how to access the parameters and return values.

I think we better move this sections in to `joomla-4-and-5-changes.md`, and just link it here.
To keep the page more clean, and avoid duplication on every other Events detail page.
What do you think?

The "Accessing Event Parameters / Arguments" looks good.

The "Returning Values" is very confusing :wink: 
I kind of understand what you wanted to say, but I not sure about others.
The events does not have return value at all now. Maybe can try something like:

--------

The legacy event whith is expecting the return value:
```php
function onFooBar() 
{
  // ...
  return $myValue;
}
```

The event class whith is expecting the result:
```php
function onFooBar(FooBarEvent $event): void
{
  // ...
  $event->addResult($myValue);
}
```

------------

**Parameters which was with "reference"**

Also there is events which had parameters as reference, in j5 this events got a special method to update the parameter (because referencing does not work anymore). I have tries do give a migration example at:
https://manual.joomla.org/migrations/44-50/removed-backward-incompatibility#module-event-onrendermodule-backward-compatibility
For `onContentPrepareData` https://manual.joomla.org/migrations/44-50/removed-backward-incompatibility#content-event-oncontentpreparedata-backward-compatibility

but this also better to be part of `joomla-4-and-5-changes.md`

> I'm thinking also that we should maybe change 'item' to 'subject' in the arguments as well. What do you think?

Yes, that is good idea. Maybe for arguments that changed their names can do kind of this:
- `subject` (formerly known as `$item` or `$row` parameter) ... the rest of description ...

robbiejackson commented:
> I think we better move this sections in to `joomla-4-and-5-changes.md`, and just link it here.
> To keep the page more clean, and avoid duplication on every other Events detail page.
> What do you think?

Yes, agreed! I'm going to do this. Otherwise it's going to end up unmaintainable.


> The event class which is expecting the result:
> 
> ```
> function onFooBar(FooBarEvent $event): void
> {
>   // ...
>   $event->addResult($myValue);
> }
> ```

The problem is that this works only if it's a concrete class. If it's a generic class it raises an exception. So if someone is trying to return a value to onContentAfterTitle in Joomla 4.4 then they can't use this, and simply returning a value won't work, so they have to use the generic event approach. 

I think I know how to make this clearer though. 


> **Parameters which was with "reference"**
> 
> Also there is events which had parameters as reference, in j5 this events got a special method to update the parameter (because referencing does not work anymore).

Ok, I wasn't aware of this. As the update functions are going to have different names I think it's best to document this in the detailed section, eg in the onContentPrepareData documentation. 

Will this just be done for some parameters and not for others? For example, for onContentPrepare where you can update the $subject/item?

Fedik commented:
> The problem is that this works only if it's a concrete class. If it's a generic class it raises an exception. 

Yes, you are right. 
My example was for "clean" listeners, specific for old and new. 
For "transitional" listeners, could add 3-rd example, similar to what you made.

```php
function onFooBar(Joomla\Event\Event $event): void
{
    // ...
    if ($event instanceof ResultAwareInterface) {
        $event->addResult($value);
    } else {   // use GenericEvent approach
        $result = $event->getArgument('result') ?: [];   // get the result argument from GenericEvent
        $result[] = $value;                              // add your return value into the array
        $event->setArgument('result', $result);  
    }
}
```

> Will this just be done for some parameters and not for others? For example, for onContentPrepare where you can update the $subject/item?

Yes, only for specific parameters, in specific events.

I tried to document all of them at https://manual.joomla.org/migrations/44-50/removed-backward-incompatibility/#module-event-onrendermodule-backward-compatibility
Can just copy paste from there, or give a link for reference.

The list:
```
onRenderModule
attributes

onAfterRenderModules
content

onPrepareModuleList, onAfterModuleList, onAfterCleanModuleList
modules

onCustomFieldsAfterPrepareField
value

onInstallerBeforeInstallation, onInstallerBeforeInstaller, onInstallerAfterInstaller
package

onInstallerBeforePackageDownload
url
headers

onContentBeforeValidateData
data

onContentPrepareData
data

onSubmitContact
data

onPreprocessMenuItems
items

onAfterGetMenuTypeOptions
items
```
These events should use `$event->updateFoobar($newValue);` also could use `$event->setArgument('foobar', $newValue);`.

Technicaly, you can use $event->setArgument() for any argument, but it is not recomended, because most of the event class is `AbstractImmutableEvent`. And after we get something like this https://github.com/joomla/joomla-cms/pull/39469 merged it will be not possible to do such thing for imutable arguments.

robbiejackson commented:
I'm going to go with Fedir's suggestion of putting the description in Joomla 4/5 changes page, and linking to it from each details page. Otherwise there's a lot of maintenance overhead. I've updated the explanation hopefully to make it clearer.

I've changed item to subject in the first content events, but I've finding a problem with NormaliseRequestDataEvent getting the subject. I think getSubject() might be missing.

Fedik commented on docs/building-extensions/plugins/plugin-events/content.md:
```diff
@@ -0,0 +1,161 @@
+
+- **`context`** - The context of the content being passed to the plugin. This is the component name and view - or name of module (e.g. com_content.article, com_contact.contact, com_users.user). Use this to check whether you are in the desired context for the plugin.
+
+- **`&subject`** (**`&item`** is also acceptable) - A reference to the item which is being rendered by the view, for example, an article, contact or user. 
```
no no, sorry, this will not work :wink: 
What I meant, that `item` is a previous name of the parameter, and use of `item` does not work with event class anymore.
this will work: `$event->getArgument('subject')`;
this will return `null`:  `$event->getArgument('item')`;

robbiejackson replied:
I checked and $event->getArgument('item') returns the item ok.

I'm finding this quite confusing though. For NormaliseRequestDataEvent you have

    $event->getArgument('context');
    $event->getArgument('data');
    $event->getArgument('subject');
    $event->getArgument('form');

And also 

    $event->getContext();
    $event->getData();
    $event->getForm();

but you can't use 

    $event->getSubject();

I think the guidance should be to use 

    $event->getContext();

instead of 

    $event->getArgument('context');

And then in the list of event arguments we should use the names which correspond to those functions which exist (ie use 'form' instead of 'subject').

There's a couple of other reasons for this:
- people will understand what 'form' means more readily than 'subject'
- it matches what's described in the API documents https://api.joomla.org/cms-5/classes/Joomla-CMS-Event-Model-NormaliseRequestDataEvent.html 

Do you foresee any problem with this approach?

Fedik replied:
> I checked and $event->getArgument('item') returns the item ok.

hm, you are right, I just checked and this happen due to legacy code handling https://github.com/joomla/joomla-cms/blob/46c3bc031670aa0550ebbc7e738d73ab3c349782/libraries/src/Event/AbstractEvent.php#L170-L182
and will stop working at some point in future.
So, it works by accident :smile: 

> I'm finding this quite confusing ...

Yes, it also was for me. And It probably will be confusing for everyone :smile: 

It comes from https://github.com/joomla/joomla-cms/blob/46c3bc031670aa0550ebbc7e738d73ab3c349782/libraries/src/Event/AbstractEvent.php#L45-L46
And when I made event classes, I took the "main" parameter and name it as a `subject` argument. To follow that description.

Later I thought it much nicer to have the getters with typehint like `$event->getFoobar()` than use of `$event->getArgument('foobar')` or `$event['foobar']`. And added these getters, but without use of `getSubject()` for `subject` argument. 
We probably can add this method at some point, to reslove this confusion. Not sure currently.

 
> I think the guidance should be to use `$event->getContext();` instead of `$event->getArgument('context');`

Yes and no. If we talking about specific event, that have this method implemented, then yes. If we talking about generic event class, that does not have this method, then no.
I mean, these methods is kind of helpers, which provides type hinting.

robbiejackson replied:
Yes, just for specific events. I think we focus on the getters in the documentation, for several reasons.

- 'subject' means different things in different events, so it's not as clear as 'form' or 'item'
- type-hinting available
- that's what the API docs show
- matches the "update" setters.

I'll update the documentation - again ;-)

Fedik commented:
> I've finding a problem with NormaliseRequestDataEvent getting the subject. I think getSubject() might be missing.

`NormaliseRequestDataEvent` is the same as `BeforeValidateDataEvent`, and child of `FormEvent`.
`subject` here is the form object.
For these events it is like: `$form = $event->getArgument('subject')` or `$form = $event->getForm()`

Fedik commented:
Please remove all `&` in the argument names, like `&foobar` to just `foobar`. It does not need anymore, and can be confusing.
Also I would probably remove it from description, from eg:
`A reference to the item which is being rendered by the view ...`
to:
`The item which is being rendered by the view ...`

But just removing `&` should be enough.


The rest is looks good now.

robbiejackson commented:
@Fedik I've removed the & symbols and changed some of the text as you suggested. Many thanks for your help sorting out this page - it was important to get it in a good state as a pattern for the rest.

Would you mind having a quick look at the updated index page to make sure it's accurate? (I've changed it a bit).
http://pr-291.manual.joomlacode.org/docs/next/building-extensions/plugins/plugin-events/

Fedik commented:
Sorry a bit late. 
I just noticed one litle thing :smile: 

The order of events is mixed:
```
onContentBeforeValidateData	Before POST data validation
onContentNormaliseRequestData	After POST data validation
```
Actualy the `onContentNormaliseRequestData` event goes first ([trigrerred in the controller](https://github.com/joomla/joomla-cms/blob/aa475c1a66370bb400c02f44c0c2830d567f4f7d/libraries/src/MVC/Controller/FormController.php#L627), before validation) , then `onContentBeforeValidateData` ([trigered in model](https://github.com/joomla/joomla-cms/blob/aa475c1a66370bb400c02f44c0c2830d567f4f7d/libraries/src/MVC/Model/FormModel.php#L213), before validation)

Also in the events descriptions it mixed, for `onContentNormaliseRequestData` it says:
```
It is triggered after the submitted form data 
(usually an array sent by the browser within a jform HTTP POST parameter) 
has been filtered and validated.
```
However it is triggered before validation.

robbiejackson commented:
Ah thanks @Fedik , I missed that. I'll fix it up in the next pull request.
